spark/src/main/scala/io/delta/sharing/spark/filters/JsonPredicates.scala

```diff
@@ -0,0 +1,346 @@
+
+  // Literals are always non-null.
+  override def isNull(ctx: EvalContext): Boolean = false
+}
+
+/**
+ * Represents a null operation on its child.
+ */
+case class IsNullOp(children: Seq[LeafOp]) extends NonLeafOp with UnaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = children(0).isNull(ctx)
+}
+
+/**
+ * The following ops represent comparison operations on their children.
+ * The (logically) supported operations are:
+ *   "=", "<", "<=", ">", ">="
+ *
+ * We only implement "=" and "<", and represent all others in terms
+ * of these two.
+ *
+ * @param children Expected size is 2.
+ */
+
+case class EqualOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = EvalHelper.equal(children, ctx)
+}
+
+case class LessThanOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = EvalHelper.lessThan(children, ctx)
+}
+
+case class LessThanOrEqualOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any =
+    EvalHelper.lessThan(children, ctx) || EvalHelper.equal(children, ctx)
+}
+
+case class GreaterThanOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any =
+    !EvalHelper.lessThan(children, ctx) && !EvalHelper.equal(children, ctx)
+}
+
+case class GreaterThanOrEqualOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = !EvalHelper.lessThan(children, ctx)
+}
+
+/**
+ * The following ops represent boolean logic related operations.
+ * The (logically) supported operations are:
+ *   "&&", "||", "!"
+ */
+
+case class AndOp(val children: Seq[NonLeafOp]) extends NonLeafOp with NaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = {
+    children.forall(c => c.evalExpectBoolean(ctx))
+  }
+}
+
+case class OrOp(val children: Seq[NonLeafOp]) extends NonLeafOp with NaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = {
+    children.exists(c => c.evalExpectBoolean(ctx))
+  }
+}
+
+case class NotOp(val children: Seq[NonLeafOp]) extends NonLeafOp with UnaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = !children(0).evalExpectBoolean(ctx)
+}
+
+// A helper for evaluating opertions.
+object EvalHelper {
+
+  // Implements "equal" between two leaf operations.
+  def equal(children: Seq[LeafOp], ctx: EvalContext): Boolean = {
```

Comment:
This only accepts leaf node, it means we don't support expressions like `(a EQ b) EQ True`?

Reply:
Your example can be expressed as: (a eq b) && True.

We only need equality for leaf nodes (columns and literals) for both partition and stats filtering as of now.
I'm intentionally starting out with a restricted version, and we can relax this as needed.